src/vm/i386/gmsx86.cpp

```diff
@@ -1295,10 +1295,10 @@ void LazyMachState::unwindLazyState(LazyMachState* baseState,
     ctx.Esi = lazyState->_esi = baseState->_esi;
     ctx.Ebx = lazyState->_ebx = baseState->_ebx;
 
-    nonVolRegPtrs.Edi = &(lazyState->_edi);
-    nonVolRegPtrs.Esi = &(lazyState->_esi);
-    nonVolRegPtrs.Ebx = &(lazyState->_ebx);
-    nonVolRegPtrs.Ebp = &(lazyState->_ebp);
+    nonVolRegPtrs.Edi = &(baseState->_edi);
```

Comment:
@janvorli Hmm... I got confused. Could you let me know the role of ``LazyMachState::unwindLazyState``? I guessed that it will recover the just before calling unmanaged function. 

I guess that the fixup that you mentioned also fixes this issue: ``ctx`` is correct after``PAL_VirtualUnwind`` (unlike nonVolRegPtrs), and the fixup will overwrite (incorrect) interior pointers (untouched by ``PAL_VirtualUnwind``) as pointer that point corresponding variables inside ``MachState`` itself that are correct.

For me, however, it seems that the fixup inside ``LazyMachState::setLazyStateFromUnwind`` is unnecessary if we properly initializes ``nonVolRegPtrs``.

As I understand, ``baseState`` is the captured state at the beginning of unmanaged function. If unmanaged fuction does not touch a callee-saved register, the corresponding value in ``baseState`` should be the one that we want to find. If unmanaged fuction touches this register, then managed code should save its previous value on the stack, ``PAL_VirtualUnwind`` should update the corresponding pointer properly.